src/vec/mod.rs

```diff
@@ -11,18 +11,88 @@ use core::{
+    /// and two type aliases for convenience:
+    ///
+    /// - [`Vec<T, N>`](crate::vec::Vec) = `VecInner<T, OwnedStorage<T, N>>`
+    /// - [`VecView<T>`](crate::vec::VecView) = `VecInner<T, ViewStorage<T>>`
+    ///
+    /// `Vec` can be unsized into `VecView`, either by unsizing coercions such as `&mut Vec -> &mut VecView` or
+    /// `Box<Vec> -> Box<VecView>`, or explicitly with [`.as_view()`](crate::vec::Vec::as_view) or [`.as_mut_view()`](crate::vec::Vec::as_mut_view).
+    ///
+    /// This trait is sealed, so you cannot implement it for your own types. You can only use
+    /// the implementations provided by this crate.
+    ///
+    /// [`VecInner`]: super::VecInner
+    /// [`Vec`]: super::Vec
+    /// [`VecView`]: super::VecView
+    #[allow(private_bounds)]
+    pub trait VecStorage<T>: VecSealedStorage<T> {}
+
+    pub trait VecSealedStorage<T> {
+        // part of the sealed trait so that no trait is publicly implemented by `OwnedVecStorage` besides `Storage`
+        fn borrow(&self) -> &[MaybeUninit<T>];
+        fn borrow_mut(&mut self) -> &mut [MaybeUninit<T>];
+    }
+
+    // One sealed layer of indirection to hide the internal details (The MaybeUninit).
+    pub struct VecStorageInner<T: ?Sized> {
+        pub(crate) buffer: T,
+    }
+
+    /// Implementation of [`VecStorage`] that stores the data in an array `[T; N]` whose size is known at compile time.
+    pub type OwnedVecStorage<T, const N: usize> = VecStorageInner<[MaybeUninit<T>; N]>;
+    /// Implementation of [`VecStorage`] that stores the data in an unsized `[T]`.
+    pub type ViewVecStorage<T> = VecStorageInner<[MaybeUninit<T>]>;
+
+    impl<T, const N: usize> VecSealedStorage<T> for OwnedVecStorage<T, N> {
+        fn borrow(&self) -> &[MaybeUninit<T>] {
+            &self.buffer
+        }
+        fn borrow_mut(&mut self) -> &mut [MaybeUninit<T>] {
+            &mut self.buffer
+        }
+    }
+    impl<T, const N: usize> VecStorage<T> for OwnedVecStorage<T, N> {}
+
+    impl<T> VecSealedStorage<T> for ViewVecStorage<T> {
+        fn borrow(&self) -> &[MaybeUninit<T>] {
+            &self.buffer
+        }
+        fn borrow_mut(&mut self) -> &mut [MaybeUninit<T>] {
+            &mut self.buffer
+        }
+    }
+    impl<T> VecStorage<T> for ViewVecStorage<T> {}
+}
+pub use storage::{OwnedVecStorage, VecStorage, ViewVecStorage};
+
+pub(crate) use storage::VecStorageInner;
+
 pub use drain::Drain;
 
-/// Base struct for [`Vec`] and [`VecView`], generic over the [`Storage`].
+/// Base struct for [`Vec`] and [`VecView`], generic over the [`VecStorage`].
 ///
 /// In most cases you should use [`Vec`] or [`VecView`] directly. Only use this
 /// struct if you want to write code that's generic over both.
-pub struct VecInner<T, S: Storage> {
+pub struct VecInner<T, S: VecStorage<T> + ?Sized> {
+    // This phantomdata is required because otherwise rustc thinks that `T` is not used
```

Comment:
Minor nit: Is this comment helpful? After all, that's usually the reason to use PhantomData.

Reply:
Indeed, it was not obvious to me initially that it would be required because `T` is used in the generic bound, but the comment does not do anything to explain why it's needed.

I removed it. Thanks.